### PR TITLE
add waitForRender method

### PR DESCRIPTION
**Pull request by jefleponot**
Hi everybody,

#1722

With CasperJs, you can take whole or partial screen-shots (with "casper.captureX" methods), but ...

Issues :

    how to be sure that rendering is what you expect (Image / Canvas / Flash / Video / SVG / Complex CSS) ?
    how to wait for the end of a css3 or JavaScript animation ?
    how to click on a screen zone without any locator (canvas / Image Area ) or dynamic selector ?

I would like to propose a new method that could compare an image with the browser's screen and wait until it matches : waitForRender

### CODE REVIEW

**Comment by mlalpho** [comment]
Excellent feature! Great job @jefleponot

**Comment by jefleponot** [comment]
Thanks @mlalpho 

Everything seems to be ok, could anyone review my code ?

@mickaelandrieu , I will write a more examples in documentation, if you are interested by this feature.

**Comment by jefleponot** [comment]
I would like to add this feature in next version of CasperJS, could someone take time to review the code ?
@mickaelandrieu , @mlalpho , @BIGjuevos

**Comment by mickaelandrieu** [comment]
I'll be available after 10th of december.

I'm thinking... we should encourage modules creation over new core contributions: what do you think @jefleponot  ?

**Comment by jefleponot** [comment]
Casperjs is a project full of possibilities (web application tests, supervision, data scraping ...). 

It needs to evolve and renew itself. But there is less and less contributions.

How to encourage contributions ? 
Perhaps, in proposing, some features that similar projects can't done. it is the goal of this PR.
As I'm not yet using NodeJS and all its environment so it's difficult for me to split CasperJS features into different modules.

It could be clever to converge towards a full asynchronous running (like @dritter proposals) to prepare the future versions of PhantomJS (latest full asynchronous webkit engine) => CasperJS 2.0.0

Friendly
jef

**Comment by mlalpho** [comment]
@jefleponot, I'd be willing to contribute some more time here. I think modularity is key to allowing diverse extensibility. 

There are still a few unknowns for me when it comes to contributing and I think it would be helpful for others too to know what all it takes to get things built and released properly. The less friction there is to get contributions accepted and released, the more willing someone would be to contribute. 

This project does have immense possibilities and I look forward to its future.

Ultimately, communication and consistency are key. We need to identify and eliminate the barriers that are stifling the potential for innovation of this project.

**Comment by mathiasm74** [comment]
@jefleponot wrote:
> How to encourage contributions?

If you want to encourage contributions then surely the first step is to make sure you don't have 40+ pull requests just sitting there, no? Am i naive?

**Comment by jefleponot** [comment]
@mathiasm74 , thanks for your proposal, could you please tell me the PR you would merge ?

File: tests/suites/casper/capture.js [hunk 1]
```diff
@@ -95,3 +95,20 @@ casper.test.begin('Casper.captureBase64() tests', 4, function(test) {
...
 });
+
+
+casper.test.begin('Casper.waitForRender() tests', 4, function(test) {
+    casper.start('tests/site/index.html', function() {
+        test.assert(this.captureBase64('png').length > 0,
+            'Casper.captureBase64() rendered a page capture as base64');
+        var base64 = "data:image/png;base64," + this.captureBase64('png', {top: 10, left: 20, width: 30, height: 40});
+        
+        casper.waitForRender(base64, function(coords ) {
```
Comment (mickaelandrieu) [inline tests/suites/casper/capture.js]:
```diff
- function(coords )
+ function(coords)
```

File: modules/clientutils.js [hunk 2]
```diff
@@ -68,6 +69,54 @@
                     });
         }
 
+        function image_difference(imageData, patternData, coords, delta) {
```
Comment (mickaelandrieu) [inline modules/clientutils.js]: if this function is about to return a boolean, how about prefix it by "is"? something like ``is_image_different`` sounds better to me :)

**Comment by jefleponot** [comment]
Hi @mickaelandrieu,

I have time today to works on CasperJS.
I have passed my mono compilation at last.
I have also rewrite this PR but I don't commit yet. It is now more mature.

## Only one issue 
template matching need a lot of resources so it could be nice to not freeze client page.
Are you bothered if I code it with "Web Workers" calls ?

Regards

**Comment by mickaelandrieu** [comment]
@jefleponot depends ... is it supported by phantomjs 1.9.8 and slimerjs 0.9.6? Or is it fine to use Web workers only if they are available ?

**Comment by kensoh** [comment]
Bumping up this thread :)

**Comment by kensoh** [comment]
Archiving PR using **ARCHIVE** label as CasperJS, like PhantomJS, is not actively maintained for now. These PRs are from 1-5 years ago but there just hasn't been maintainers to continue the work.

**Pull request closed without merging**